SudokuSolver/SudokuSolver.1
.\"Modified from man(1) of FreeBSD, the NetBSD mdoc.template, and mdoc.samples.
.\"See Also:
.\"man mdoc.samples for a complete listing of options
.\"man mdoc for the short list of editing options
.\"/usr/share/misc/mdoc.template
.Dd 17/04/14               \" DATE 
.Dt SudokuSolver 1      \" Program name and manual section number 
.Os Darwin
.Sh NAME                 \" Section Header - required - don't modify 
.Nm SudokuSolver,
.\" The following lines are read in generating the apropos(man -k) database. Use only key
.\" words here as the database is built based on the words here and in the .ND line. 
.Nm Other_name_for_same_program(),
.Nm Yet another name for the same program.
.\" Use .Nm macro to designate other names for the documented program.
.Nd This line parsed for whatis database.
.Sh SYNOPSIS             \" Section Header - required - don't modify
.Nm
.Op Fl abcd              \" [-abcd]
.Op Fl a Ar path         \" [-a path] 
.Op Ar file              \" [file]
.Op Ar                   \" [file ...]
.Ar arg0                 \" Underlined argument - use .Ar anywhere to underline
arg2 ...                 \" Arguments
.Sh DESCRIPTION          \" Section Header - required - don't modify
Use the .Nm macro to refer to your program throughout the man page like such:
.Nm
Underlining is accomplished with the .Ar macro like this:
.Ar underlined text .
.Pp                      \" Inserts a space
A list of items with descriptions:
.Bl -tag -width -indent  \" Begins a tagged list 
.It item a               \" Each item preceded by .It macro
Description of item a
.It item b
Description of item b
.El                      \" Ends the list
.Pp
A list of flags and their descriptions:
.Bl -tag -width -indent  \" Differs from above in tag removed 
.It Fl a                 \"-a flag as a list item
Description of -a flag
.It Fl b
Description of -b flag
.El                      \" Ends the list
.Pp
.\" .Sh ENVIRONMENT      \" May not be needed
.\" .Bl -tag -width "ENV_VAR_1" -indent \" ENV_VAR_1 is width of the string ENV_VAR_1
.\" .It Ev ENV_VAR_1
.\" Description of ENV_VAR_1
.\" .It Ev ENV_VAR_2
.\" Description of ENV_VAR_2
.\" .El                      
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag -width "/Users/joeuser/Library/really_long_file_name" -compact
.It Pa /usr/share/file_name
FILE_1 description
.It Pa /Users/joeuser/Library/really_long_file_name
FILE_2 description
.El                      \" Ends the list
.\" .Sh DIAGNOSTICS       \" May not be needed
.\" .Bl -diag
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .It Diagnostic Tag
.\" Diagnostic informtion here.
.\" .El
.Sh SEE ALSO 
.\" List links in ascending order by section, alphabetically within a section.
.\" Please do not reference files that do not exist without filing a bug report
.Xr a 1 , 
.Xr b 1 ,
.Xr c 1 ,
.Xr a 2 ,
.Xr b 2 ,
.Xr a 3 ,
.Xr b 3 
.\" .Sh BUGS              \" Document known, unremedied bugs 
.\" .Sh HISTORY           \" Document history if command behaves in a unique manner

SudokuSolver/SudokuSolver.m
//
//  SudokuSolver.m
//  SudokuSolver
//
//  Created by walterfernandes on 17/04/14.
//  Copyright (c) 2014 walterfernandes. All rights reserved.
//

#import "SudokuSolver.h"
#import "SudokuSlot.h"

@implementation SudokuSolver

-(id)init{
    if ((self = [super init])){
        end = CFAbsoluteTimeGetCurrent();
        start = CFAbsoluteTimeGetCurrent();
        totalTime = 0.0f;
    }
    
    return self;
}

-(void)printGameInstance:(NSArray*)gameInstance {
    
    printf("\n|");
    for (int j = 0; j < kGameSize; j++) {
        printf("----");
        if (j%3 == 2)
            printf("|");
    }
    printf("\n|");
    
    for (int i = 0; i < kGameSize; i++){
        for (int j = 0; j < kGameSize; j++){
            printf(" %2d ", [gameInstance[i][j] intValue]);
            if (j%3 == 2)
                printf("|");
        }
        
        printf("\n|");
        if (i%3 == 2){
            for (int j = 0; j < kGameSize; j++) {
                printf("----");
                if (j%3 == 2)
                    printf("|");
            }
            if (i < (kGameSize-1))
                printf("\n|");
        }
    }
    printf("\n");
}

-(NSArray*)getAllSlots:(NSArray*)gameInstace{
    NSMutableArray *slots = [[NSMutableArray alloc] initWithCapacity:kGameSize^2];
    
    for (int x = 0; x < kGameSize; x++){
        for (int y = 0; y < kGameSize; y++){
            if ([gameInstace[x][y] intValue] == 0) {
                SudokuSlot *slot = [SudokuSlot new];
                slot.p = MakePosition(x, y);
                slot.f = [self getFieldScore:slot.p forGame:gameInstace];
                slot.possibles = [self getPossibles:slot.p forGameInstace:gameInstace];
                [slots addObject:slot];
            }
        }
    }

    return slots;
}

-(SudokuSlot*)getFirstSlots:(NSArray*)gameInstace{
    for (int i = 0; i < kGameSize; i++){
        for (int j = 0; j < kGameSize; j++){
            if ([gameInstace[i][j] intValue] == 0) {
                SudokuSlot *slot = [SudokuSlot new];
                slot.p = MakePosition(i, j);
                slot.f = [self getFieldScore:slot.p forGame:gameInstace];
                slot.possibles = [self getPossibles:slot.p forGameInstace:gameInstace];
                return slot;
            }
        }
    }
    return nil;
}

-(BOOL)isFinishGame:(NSArray*)gameInstace{
    BOOL line[kGameSize];
    BOOL column[kGameSize];
    BOOL square[kGameSize];
    
    //verifica repetição nas linhas e colunas
    for (int i = 0; i < kGameSize; i++){
        
        //reseta vetor de controle
        for (int x = 0; x < kGameSize; x++)
            line[x] = NO, column[x] = NO;
        
        for (int j = 0; j < kGameSize; j++){
            int l = [gameInstace[i][j] intValue] - 1;
            int c = [gameInstace[j][i] intValue] - 1;
            
            if ([gameInstace[i][j] intValue] == 0 || line[l] || column[c]) return NO;
            
            line[l] = YES;
            column[c] = YES;
            
        }
    }
    
    //verifica repetição no quadrante
    for (int k = 0; k < kGameSize; k++){
        
        ///reseta vetor de controle
        for (int x = 0; x < kGameSize; x++)
            square[x] = NO;
        
        for (int i = 0; i < kGameSize/3; i++){
            for (int j = 0; j < kGameSize/3; j++){
                int l = ([gameInstace[i + k%3*3][j + k%3*3] intValue] - 1);
                
                if (square[l]) return NO;
                
                square[l] = YES;
            }
        }
    }
    
    return YES;
}

-(int)getFieldScore:(SudokuPosition)p forGame:(NSArray*)gameInstance
{
    int fieldScore = 0;
    for( int k = 0; k < 9; k++ )
    {
        if( ( k != p.i && gameInstance[k][p.j] == gameInstance[p.i][p.j] ) ||
           ( k != p.j && gameInstance[p.i][k] == gameInstance[p.i][p.j] ) )
            ++fieldScore;
    }
    int sq_dim = kGameSize / 3;
    int row = p.i / sq_dim;
    int col = p.j / sq_dim;
    int row_from, row_to, col_from, col_to;
    
    row_from = row*sq_dim;
    row_to = row_from + sq_dim - 1;
    col_from = col*sq_dim;
    col_to = col_from + sq_dim - 1;
    for( int k = row_from; k <= row_to; k++ )
        for ( int l = col_from; l <= col_to; l++ )
            if( k != p.i && l != p.j && gameInstance[k][l] == gameInstance[p.i][p.j])
                ++fieldScore;
    
    return fieldScore;
}

-(NSArray*)getPossibles:(SudokuPosition)p forGameInstace:(NSArray*)gameInstance{
    NSMutableArray *possibles = [NSMutableArray arrayWithArray:@[@(1), @(2), @(3), @(4), @(5), @(6), @(7), @(8), @(9)]];
    
    for (int i = 0; i < kGameSize; i++){
        [possibles removeObject:gameInstance[p.i][i]];
        [possibles removeObject:gameInstance[i][p.j]];
    }
    
    
    int sq_dim = kGameSize / 3;
    int row = p.i / sq_dim;
    int col = p.j / sq_dim;
    int row_from, row_to, col_from, col_to;
    
    row_from = row*sq_dim;
    row_to = row_from + sq_dim - 1;
    col_from = col*sq_dim;
    col_to = col_from + sq_dim - 1;
    for( int k = row_from; k <= row_to; k++ )
        for ( int l = col_from; l <= col_to; l++ )
             [possibles removeObject:gameInstance[k][l]];
    
    return possibles;
}


-(NSMutableArray*)copyGameInstance:(NSArray*)gameInstance{
    NSMutableArray *copy = [NSMutableArray new];
    
    for (NSArray *line in gameInstance){
        NSMutableArray *lineCopy =  [NSMutableArray new];
        
        for (NSNumber *value in line)
            [lineCopy addObject:[value copy]];
        
        [copy addObject:lineCopy];
    }
    
    return copy;
}


-(BOOL)resolveGameInstace:(NSArray*)gameInstace{
    
    numTest++;

    totalTime = end - start;
    end = CFAbsoluteTimeGetCurrent();
    
    if ([self isFinishGame:gameInstace])
        return YES;
    else {
        SudokuSlot *slot = [self getFirstSlots:gameInstace];
        if (slot)
            for (NSNumber *value in slot.possibles){
                NSMutableArray *newGameInstace = [self copyGameInstance:gameInstace];
                newGameInstace[slot.p.i][slot.p.j] = value;
                if ([self resolveGameInstace:newGameInstace])
                    return YES;
            }
        
        return NO;
    }
}

-(float)getTotalTime{
    return totalTime;
}

-(int)getNumTest{
    return numTest;
}

@end
